Minor documentation/API improvements.

Lymia opened the pull request.

Lymia committed:
Improve documentation for all types.

Lymia committed:
Make the order of `MemoryTypeId` parameters consistant.

Lymia committed:
Implement used/allocated functions.

zakarumych commented on src/chunked.rs:
```diff
@@ -215,20 +229,22 @@ impl<T> ChunkedAllocator<T> {
+    ///                       allocator.
+    /// - `min_block_size`: The minimum block size used by this allocator in bytes. Allocations
+    ///                     significantly than this may incur much larger overhead.
+    /// - `max_chunk_size`: The maximum size of chunks allocated from the underlying allocator
```
blocks larger than `max_chunk_size / blocks_per_chunk` can't be allocated.

Lymia replied:
You *sure*? Reading through the code, it is perfectly willing to allocate chunks with less than `blocks_per_chunk` blocks, when running up against `max_chunk_size`

zakarumych replied:
Uh. I forgot. Sorry 😏

Lymia replied:
Actually, with that in mind, `CombinedAllocator` should probably allocate in `root` rather than `combined` when `reqs.size > self.chunks.max_chunk_size() / 2` rather than when `reqs.size > self.chunks.max_chunk_size()`.

(Since it'd take up the whole chunk then, anyways)

zakarumych commented on src/combined.rs:
```diff
@@ -96,34 +107,37 @@ where
-                .map(|ArenaBlock(block, tag)| CombinedBlock(block, CombinedTag::Arena(tag))),
+                .map(|ArenaBlock(block, tag)| CombinedBlock(block, CombinedTag::Arena(tag)))?,
             Type::General => {
                 if reqs.size > self.chunks.max_chunk_size() {
```
This could be wrong.

Lymia replied:
How so? This is a simple mechanical transformation, moving a `?` from outside the `match` into the match arms.

zakarumych replied:
Not the change. I was referring to the `if reqs.size > ...` line.
But as you reminded to me that it indeed can allocate blocks with size up to `max_chunk_size` this line seems to me correct now.

zakarumych commented on src/lib.rs:
```diff
@@ -160,7 +160,7 @@ pub trait MemoryAllocator<B: Backend>: Debug {
 }
 
-/// Trait that allows to sub-allocate memory blocks from another allocator.
+/// Trait for allocators that memory blocks from another allocator it doesn't own.
```
Missing verb

zakarumych commented on src/combined.rs:
```diff
@@ -96,34 +107,37 @@ where
                         .alloc(device, (), reqs)
-                        .map(|block| CombinedBlock(block, CombinedTag::Root))
+                        .map(|block| CombinedBlock(block, CombinedTag::Root))?;
+                    self.root_used += block.size();
```
This value can be fetched from root allocator itself.

Lymia replied:
This will double-count the allocations in `RootAllocator` from the `ArenaAllocator` or `ChunkedAllocator`

zakarumych replied:
Oh. Sure. You're right.

zakarumych commented on src/root.rs:
```diff
@@ -17,6 +17,7 @@ pub struct RootAllocator<B> {
     relevant: Relevant,
     id: MemoryTypeId,
     allocations: usize,
+    used: u64,
```
This makes `allocations` obsolete.

zakarumych reviewed (changes requested):
This is nice. Only few nits

Lymia committed:
Change `RootAllocator` to only use `used` to track `is_used`.

Lymia committed:
Fix grammar error.

zakarumych reviewed (approved):
Great, thanks!

Lymia committed:
Allocate from `root` rather than `chunked` in CombinedAllocator earlier.

Since when the size of the block > `self.chunks.max_chunk_size() / 2`, `ChunkedAllocator` dedicates the whole allocation to the block anyway, this wastes less memory.
